File: Sample/Global/Components/PathPreviewComponent.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import TYPE_CHECKING
import Engine
from Engine import RectangleShape, Vector2f, Color
from .ComponentBase import ComponentBase
from .PathRouteState import PathRouteState

if TYPE_CHECKING:
    from ..GameMap import GameMap
    from ..Camera import Camera


class PathPreviewComponent(ComponentBase):
    r"""
    \brief Component that renders a visual preview of the pathfinding route.

    This component renders semi-transparent rectangles along the
    planned path to show where the actor will move.
    """

    def __init__(self, gameMap: GameMap, routeState: PathRouteState) -> None:
        r"""
        \brief Initialize the PathPreviewComponent.

        - gameMap: The game map this component operates on.
        - routeState: The path route state to read the planned path from.
        """
        super().__init__(gameMap)
        self._routeState = routeState
        self._fillColor = Color(80, 180, 255, 110)
        self._outlineColor = Color(120, 220, 255, 180)

    def onRender(self, camera: Camera) -> None:
        r"""
        \brief Render the path preview on the map.

        - camera: The camera to use for rendering.
        """
        route = self._routeState.getRoute()
        if len(route) == 0:
            return
        cellSize = Engine.CellSize
        pad = max(1.0, cellSize * 0.12)
        size = max(1.0, cellSize - pad * 2.0)
        for cell in route:
            rect = RectangleShape(Vector2f(size, size))
            rect.setPosition(Vector2f(cell.x * cellSize + pad, cell.y * cellSize + pad))
            rect.setFillColor(self._fillColor)
            rect.setOutlineColor(self._outlineColor)
            rect.setOutlineThickness(1.0)
            camera.render(rect)

File: Sample/Scripts/Global/Components/PathPreviewComponent.lua
local Engine = require("Engine")
local ComponentBase = require("Global.Components.ComponentBase")
local Pool = require("Global.Pool")

local PathPreviewComponent = {}

function PathPreviewComponent:init(gameMap, routeState)
    super(PathPreviewComponent, self).init(gameMap)
    ---@type PathRouteState
    self._routeState = routeState
    self._fillColour = sf.Color.new(80, 180, 255, 110)
    self._outlineColour = sf.Color.new(120, 220, 255, 180)
    self._padding = math.max(1.0, Engine.CellSize * 0.12)
    local size = math.max(1.0, Engine.CellSize - self._padding * 2.0)
    self._rectangleSize = sf.Vector2f.new(size, size)
    self._rectangle = sf.RectangleShape.new(self._rectangleSize)
    self._rectangle:setFillColor(self._fillColour)
    self._rectangle:setOutlineColor(self._outlineColour)
    self._rectangle:setOutlineThickness(1.0)
    self._cachedRouteRevision = -1
    ---@type sf.Vector2i[]
    self._cachedRoute = {}
end

function PathPreviewComponent:onRender(camera)
    local routeRevision = self._routeState:getRevision()
    if routeRevision ~= self._cachedRouteRevision then
        self._cachedRoute = self._routeState:getRoute()
        self._cachedRouteRevision = routeRevision
    end
    if not bool(self._cachedRoute) then
        return
    end
    local cellSize = Engine.CellSize
    ---@diagnostic disable: need-check-nil
    local position = Pool.Get("sf.Vector2f", sf.Vector2f, {
        x = self._cachedRoute[1].x * cellSize + self._padding,
        y = self._cachedRoute[1].y * cellSize + self._padding
    })
    ---@diagnostic enable: need-check-nil
    for index, cell in ipairs(self._cachedRoute) do
        if index > 1 then
            position.x = cell.x * cellSize + self._padding
            position.y = cell.y * cellSize + self._padding
        end
        self._rectangle:setPosition(position)
        camera:render(self._rectangle)
    end
    Pool.Put("sf.Vector2f", position)
end

return class(PathPreviewComponent, ComponentBase)
